astropy/table/info.py

```diff
@@ -0,0 +1,128 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import sys
+import os
+
+import numpy as np
+from ..extern import six
+
+__all__ = ['table_info', 'TableInfo']
+
+def table_info(tbl, option='attributes', out=''):
+    """
+    Write summary information about column to the ``out`` filehandle.
+    By default this prints to standard output via sys.stdout.
+
+    The ``option`` argument specifies what type of information
+    to include.  This can be a string, a function, or a list of
+    strings or functions.  Built-in options are:
+
+    - ``attributes``: basic column meta data like ``dtype`` or ``format``
+    - ``stats``: basic statistics: minimum, mean, and maximum
+
+    If a function is specified then that function will be called with the
+    column as its single argument.  The function must return an OrderedDict
+    containing the information attributes.
+
+    If a list is provided then the information attributes will be
+    appended for each of the options, in order.
+
+    Examples
+    --------
+    >>> from astropy.table.table_helpers import simple_table
+    >>> t = simple_table(size=2, kinds='if')
+    >>> t['a'].unit = 'm'
+    >>> t.info()
+    <Table length=2>
+    name  dtype  unit
+    ---- ------- ----
+       a   int32    m
+       b float32
+
+    >>> t.info('stats')
+    <Table length=2>
+    name mean std min max
+    ---- ---- --- --- ---
+       a  1.5 0.5   1   2
+       b  1.5 0.5 1.0 2.0
+
+    Parameters
+    ----------
+    option: str, function, list of (str or function)
+        Info option (default='attributes')
+    out: file-like object, None
+        Output destination (default=sys.stdout).  If None then a
+        Table with information attributes is returned
+
+    Returns
+    -------
+    info: `~astropy.table.Table` if out==None else None
+    """
+    from .table import Table
+
+    if out == '':
+        out = sys.stdout
+
+    descr_vals = [tbl.__class__.__name__]
+    if tbl.masked:
+        descr_vals.append('masked=True')
+    descr_vals.append('length={0}'.format(len(tbl)))
+
+    outlines = ['<' + ' '.join(descr_vals) + '>']
+
+    cols = tbl.columns.values()
+    if tbl.colnames:
+        infos = []
+        for col in cols:
+            infos.append(col.info(option, out=None))
+
+        info = Table(infos, names=list(infos[0]))
+    else:
+        info = Table()
+
+    if out is None:
+        return info
+
+    # Since info is going to a filehandle for viewing then remove uninteresting
+    # columns.
+    for name in info.colnames:
+        if np.all(info[name] == ''):
+            del info[name]
+
+    if 'class' in info.colnames:
+        # Remove 'class' info column if all table columns are the same class
+        # and they are not mixin columns.
+        uniq_types = set(type(col) for col in cols)
+        if len(uniq_types) == 1 and not tbl._is_mixin_column(cols[0]):
```

Comment:
Here is a logically equivalent version that is less obscure, but still relies on the fact that one *cannot* have a Quantity-mixin in a `Table` (it will always be converted to `Column`):
```
    if 'class' in info.colnames:
        # Remove 'class' info column if all table columns are the same class
        # and they are either the default class for that table or else Quantity.
        uniq_types = set(type(col) for col in cols)
        if len(uniq_types) == 1 and isinstance(cols[0], (tbl.ColumnClass, u.Quantity)):
            del info['class']
```
When I wrote this I started thinking "why is Quantity so special?".  It's just another mixin class, and has some of the same failings that other mixins have, most notably lack of masking.  The fact that there is a QTable is only to differentiate what to do with a thingy-with-units when added to a table (Quantity or Column).  

But in the end Quantity is just another mixin along with the rest, and I think it's important to always highlight this distinction by explicitly showing the class just as we would for a QTable full of Coordinates.

I'm inclined to go with:
```
        if len(uniq_types) == 1 and isinstance(cols[0], tbl.ColumnClass):
```
In the end if it's just horrible this is a soft API (just an informational output) so we can change it.